### PR TITLE
docs: Improve the codesigning docs

### PR DESCRIPTION
**Pull request by felixrieseberg**
#### Description of Change

This PR does a _best effort_ to improve our code-signing docs. I believe that this doc being so long may have uncovered some opportunities in our tools, but for now, this document should be good enough to be a single stop for everyone trying to sign their apps.

#### Checklist

- [x] PR description included and stakeholders cc'd
- [x] `npm test` passes
- [x] relevant documentation is changed or added
- [x] PR title follows semantic [commit guidelines](https://github.com/electron/electron/blob/master/docs/development/pull-requests.md#commit-message-guidelines)
- [x] [PR release notes](https://github.com/electron/clerk/blob/master/README.md) describe the change in a way relevant to app developers, and are [capitalized, punctuated, and past tense](https://github.com/electron/clerk/blob/master/README.md#examples).

#### Release Notes

Notes: no-notes

### CODE REVIEW

File: docs/tutorial/code-signing.md [hunk 1]
```diff
@@ -6,52 +6,154 @@ created by you.
...
+your application. You can find [its documentation
+here](https://www.electron.build/code-signing).
+
+## `electron-packager`
+
+Even if you're not using an integrated build pipeline like Forge or Builder, you
+are likely using [`electron-packager`], which includes [`electron-osx-sign`] and
+[`electron-notarize`].
+
+If you're using packager's API, you can pass [in configuration that both signs
```
Comment (malept) [inline docs/tutorial/code-signing.md]:
```suggestion
If you're using Packager's API, you can pass [in configuration that both signs
```

File: docs/tutorial/code-signing.md [hunk 2]
```diff
@@ -6,52 +6,154 @@ created by you.
...
+    'hardened-runtime': true,
+    entitlements: 'entitlements.plist',
+    'entitlements-inherit': 'entitlements.plist',
+    'signature-flags': 'library'
+  },
+  osxNotarize: {
+    appleId: 'felix@felix.fun',
+    appleIdPassword: 'my-apple-id-password'
+  }
+})
```
Comment (malept) [inline docs/tutorial/code-signing.md]: It's technically an async function but Node doesn't have top-level await yet, so to hint that it's async:

```suggestion
}).then(appPaths => console.log('Apps have been packaged at:', appPaths))
```

File: docs/tutorial/code-signing.md [hunk 3]
```diff
@@ -6,52 +6,154 @@ created by you.
...
+  <dict>
+    <key>com.apple.security.cs.allow-jit</key>
+    <true/>
+    <key>com.apple.security.cs.allow-unsigned-executable-memory</key>
+    <true/>
+    <key>com.apple.security.cs.debugger</key>
+    <true/>
+  </dict>
+</plist>
+```
```
Comment (malept) [inline docs/tutorial/code-signing.md]: 🤔 

💡 Perhaps `electron-osx-sign` should default to use these entitlements for `entitlements` / `entitlements-inherit` if they aren't set.

File: docs/tutorial/code-signing.md [hunk 4]
```diff
@@ -87,8 +194,8 @@ See the [Windows Store Guide].
...
-[`electron-winstaller`]: https://github.com/electron/windows-installer
-[Xcode]: https://developer.apple.com/xcode
-[signing certificates]: https://github.com/electron/electron-osx-sign/wiki/1.-Getting-Started#certificates
-[Mac App Store Guide]: mac-app-store-submission-guide.md
-[Windows Store Guide]: windows-store-guide.md
+[`electron-winstaller`]: https://github.com/electron/windows-installer [Xcode]:
+https://developer.apple.com/xcode [signing certificates]:
+https://github.com/electron/electron-osx-sign/wiki/1.-Getting-Started#certificates
+[Mac App Store Guide]: mac-app-store-submission-guide.md [Windows Store Guide]:
+windows-store-guide.md
```
Comment (malept) [inline docs/tutorial/code-signing.md]:
```suggestion
[`electron-winstaller`]: https://github.com/electron/windows-installer
[Xcode]: https://developer.apple.com/xcode
[signing certificates]: https://github.com/electron/electron-osx-sign/wiki/1.-Getting-Started#certificates
[Mac App Store Guide]: mac-app-store-submission-guide.md
[Windows Store Guide]: windows-store-guide.md
```

File: docs/tutorial/code-signing.md [hunk 5]
```diff
@@ -62,19 +164,24 @@ See the [Mac App Store Guide].
...
-You can get a code signing certificate from a lot of resellers. Prices vary, so it may be worth your time to shop around. Popular resellers include:
+You can get a code signing certificate from a lot of resellers. Prices vary, so
+it may be worth your time to shop around. Popular resellers include:
 
 * [digicert](https://www.digicert.com/code-signing/microsoft-authenticode.htm)
 * [Comodo](https://www.comodo.com/landing/ssl-certificate/authenticode-signature/)
 * [GoDaddy](https://au.godaddy.com/web-security/code-signing-certificate)
-* Amongst others, please shop around to find one that suits your needs, Google is your friend :)
+* Amongst others, please shop around to find one that suits your needs, Google
+  is your friend :)
```
Comment (malept) [inline docs/tutorial/code-signing.md]: Gotta have that new-school emoji

```suggestion
  is your friend 😄
```

File: docs/tutorial/code-signing.md [hunk 6]
```diff
@@ -6,52 +6,154 @@ created by you.
...
+Properly preparing macOS applications for release require two steps: First, the
+app needs to be code-signed. Then, the app needs to be uploaded to Apple for a
+process called "notarization", where automated systems will further verify that
+your app isn't doing anything to endanger its users.
+
+First, ensure that you fullfil the requirements for signing and notarizing your
+app:
 
 1. Enroll in the [Apple Developer Program] (requires an annual fee)
 2. Download and install [Xcode]
```
Comment (malept) [inline docs/tutorial/code-signing.md]:
```suggestion
2. Download and install [Xcode] - this requires a computer running macOS
```

File: docs/tutorial/code-signing.md [hunk 7]
```diff
@@ -6,52 +6,154 @@ created by you.
...
 If you are building an Electron app that you intend to package and distribute,
-it should be code-signed. The Mac and Windows app stores do not allow unsigned
-apps.
+it should be code-signed.
 
-# Signing macOS builds
+# Signing & notarizing macOS builds
 
-Before signing macOS builds, you must do the following:
+Properly preparing macOS applications for release require two steps: First, the
```
Comment (malept) [inline docs/tutorial/code-signing.md]:
```suggestion
Properly preparing macOS applications for release requires two steps: First, the
```

File: docs/tutorial/code-signing.md [hunk 8]
```diff
@@ -6,52 +6,154 @@ created by you.
...
+app needs to be code-signed. Then, the app needs to be uploaded to Apple for a
+process called "notarization", where automated systems will further verify that
+your app isn't doing anything to endanger its users.
+
+First, ensure that you fullfil the requirements for signing and notarizing your
+app:
```
Comment (malept) [inline docs/tutorial/code-signing.md]:
```suggestion
To start the process, ensure that you fulfill the requirements for signing and
notarizing your app:
```

File: docs/tutorial/code-signing.md [hunk 9]
```diff
@@ -6,52 +6,153 @@ created by you.
 On macOS the system can detect any change to the app, whether the change is
```
Comment (malept) [inline docs/tutorial/code-signing.md]:
```suggestion
On macOS, the system can detect any change to the app, whether the change is
```

File: docs/tutorial/code-signing.md [hunk 10]
```diff
@@ -6,52 +6,153 @@ created by you.
...
 introduced accidentally or by malicious code.
 
-On Windows the system assigns a trust level to your code signing certificate which
-if you don't have, or if your trust level is low will cause security dialogs to
-appear when users start using your application.  Trust level builds over time
-so it's better to start code signing as early as possible.
+On Windows the system assigns a trust level to your code signing certificate
```
Comment (malept) [inline docs/tutorial/code-signing.md]:
```suggestion
On Windows, the system assigns a trust level to your code signing certificate
```

File: docs/tutorial/code-signing.md [hunk 11]
```diff
@@ -6,52 +6,153 @@ created by you.
...
+which if you don't have, or if your trust level is low will cause security
```
Comment (malept) [inline docs/tutorial/code-signing.md]:
```suggestion
which if you don't have, or if your trust level is low, will cause security
```

File: docs/tutorial/code-signing.md [hunk 12]
```diff
@@ -6,52 +6,153 @@ created by you.
...
-
-To automate this process, you can use the [`electron-notarize`] module. You
-do not necessarily need to complete this step for every build you make – just
-the builds you intend to ship to users.
+Electron's ecosystem favors configuration and freedom, so there are multiple
+ways to get your application signed and notarized.
+
+## `electron-forge`
+
+If you're using Electron's favorite build tool, getting your application signed
```
Comment (malept) [inline docs/tutorial/code-signing.md]: > Electron's favorite build tool

😻

File: docs/tutorial/code-signing.md [hunk 13]
```diff
@@ -6,52 +6,153 @@ created by you.
...
+and notarized requires a few additions to your configuration. Forge is a
```
Comment (malept) [inline docs/tutorial/code-signing.md]:
```suggestion
and notarized requires a few additions to your configuration. [Forge](https://electronforge.io) is a
```

File: docs/tutorial/code-signing.md [hunk 14]
```diff
@@ -6,52 +6,153 @@ created by you.
...
+
+
+## `electron-builder`
+
+Electron Builder comes with a custom solution for signing your application. You
+can find [its documentation here](https://www.electron.build/code-signing).
+
+## `electron-packager`
+
+Even if you're not using an integrated build pipeline like Forge or Builder, you
```
Comment (malept) [inline docs/tutorial/code-signing.md]:
```suggestion
If you're not using an integrated build pipeline like Forge or Builder, you
```

File: docs/tutorial/code-signing.md (hunk 2 above)
Comment (felixrieseberg) [reply docs/tutorial/code-signing.md]: I feel like that's a bit outside of documentation scope here, I'd like to keep the example as small as possible

Comment (malept) [reply docs/tutorial/code-signing.md]: That's fair.

**Pull request merged**